### DISCUSSION TITLE
How do I implement fast/slow movement with the Shift-key correctly?

### DISCUSSION THREAD

**Discussion by LegionaryCohort** [AlmasB/FXGL#1312] [Q&A] [answered]
Hi,
First of all, great game engine, having a lot of fun playing around with it so far!

Context for my question:
I have a semi-static map (top-down view) and I'm trying to implement input handling that allows the player to move the viewport.
Nothing fancy, just WASD-keys to move the view around. This works great so far.
I'm now trying to add the ability to move the view around faster when pressing the Shift key as well.

So far I've tried the following:

1) **Add seperate UserActions for the keys with and without the Shift modifier.**
The UserActions simply set the viewport speed, which in turn is used in onUpdate to move the viewport around.
This mostly works as expected, however with one caveat. When I press the A key, and then press the Shift key (while still holding the A key down), it does not switch to the fast movement. Vice versa, if I start with Shift-A and then release the Shift key (while still holding the A key), the movement stops entirely for a short moment, then resumes in slow mode.
The cause appears to be that as long as the A-key remains pressed, the Shift key is not registered. Or in the second scenario, the Shift-A combination is registered and then stops and it takes the engine a short moment to register that the A-key is still being pressed.
```
private double speed = 0;

@Override
protected void initInput() {
	var moveLeft = new UserAction("Move Camera Left (Slow)") {
		@Override
		protected void onActionBegin() {
			speed = -600;
		}
		@Override
		protected void onActionEnd() {
			speed = 0;
		}
	};
	var moveLeftFast = new UserAction("Move Camera Left (Fast)") {
		@Override
		protected void onActionBegin() {
			speed = -1200;
		}
		@Override
		protected void onActionEnd() {
			speed = 0;
		}
	};
	FXGL.getInput().addAction(moveLeft, KeyCode.A);
	FXGL.getInput().addAction(moveLeftFast, KeyCode.A, InputModifier.SHIFT);
}

@Override
protected void onUpdate(double tpf) {
	FXGL.getGameScene().getViewport().setX(viewport.getX() + speed * tpf);
}
```

2) **Add separate UserActions for the WASD keys and the shift key.**
The WASD keys again set the `speed`, the shift key sets a `factor`. In onUpdate the viewport is moved according to `speed*factor`.
This simply does not work, as the engine throws an `IllegalArgumentException` when trying to bind an action to the Shift key.
If I bind `toggleSpeed` to a different (valid) key however, this gives me exactly the result I would like to achieve, just with the wrong key.
```
private double speed = 0;
private double factor = 1;

@Override
protected void initInput() {
	var toggleSpeed = new UserAction("Toggle Camera Speed") {
		@Override
		protected void onActionBegin() {
			factor = 2;
		}
		@Override
		protected void onActionEnd() {
			speed = 1;
		}
	};
	var moveLeft = new UserAction("Move Camera Left") {
		@Override
		protected void onActionBegin() {
			speed = -600;
		}
		@Override
		protected void onActionEnd() {
			speed = 0;
		}
	};
	FXGL.getInput().addAction(toggleSpeed, KeyCode.SHIFT);
	FXGL.getInput().addAction(moveLeft, KeyCode.A);
}

@Override
protected void onUpdate(double tpf) {
	FXGL.getGameScene().getViewport().setX(viewport.getX() + speed * factor * tpf);
}
```

3) **Add a TriggerListener that handles the Shift-logic internally**
The TriggerListener checks for the WASD-keys being pressed. If detected, it will check if the Shift-key is also pressed and set the viewport speed accordingly.
This does not work, `keyTrigger.getModifier()` always returns `NONE`, regardless of whether the Shift key is pressed or not.
The regular movement works just fine though, the Shift key just seems to never be registered.
(On that note, this does feel like a bug, although I'm not entirely certain it isn't just me doing something wrong).
```
private double speed = 0;
private double factor = 1;

@Override
protected void initInput() {
	FXGL.getInput().addTriggerListener(
		new TriggerListener() {
			@Override
			protected void onAction(Trigger trigger) {
				if (trigger.isKey()) {
					var keyTrigger = (KeyTrigger) trigger;
					if (keyTrigger.getKey() == KeyCode.A) {
						speed = -600;
						factor = (keyTrigger.getModifier() == InputModifier.SHIFT) ? 2 : 1;
					}
				}
			}

			@Override
			protected void onActionEnd(Trigger trigger) {
				if (trigger.isKey()) {
					var keyTrigger = (KeyTrigger) trigger;
					if (keyTrigger.getKey() == KeyCode.A) {
						speed = 0;
						factor = 1;
					}
				}
			}
		}
	);
}

@Override
protected void onUpdate(double tpf) {
	FXGL.getGameScene().getViewport().setX(viewport.getX() + speed * factor * tpf);
}
```

So, my question:
How do I implement the behaviour that I want, i.e. smooth switching between regular and fast movement, without having to release the movement keys in between?
Note that I also tried using the onAction method, rather than the onActionBegin (figuring it might simply be that the single starting tick was the problem), but this made no difference.
I feel like I'm probably missing something obvious or I'm just misusing the API, so if you could give me any pointers on this, that would be great!
(And sorry for the long post, I'm hoping the code samples clarify what I'm doing, or trying to do)

**Comment by AlmasB** [comment]
Hi,

The following could work:

1. Add a user action that moves left to A and also SHIFT+A. Both actions are identical and just move left.
2. Add a trigger listener (some kotlin code below, should be easy to adapt to java)
```
            addTriggerListener(object : TriggerListener() {
                override fun onKeyBegin(keyTrigger: KeyTrigger) {
                    if (keyTrigger.key == KeyCode.SHIFT) {
                        speedMult = 5.0
                    }
                }

                override fun onKeyEnd(keyTrigger: KeyTrigger) {
                    if (keyTrigger.key == KeyCode.SHIFT) {
                        speedMult = 1.0
                    }
                }
            })
```

Let me know if this works.

**Reply by LegionaryCohort** [reply]
Hi,
Thanks for the quick reply :)

I've tried your suggestion and it works for the basic case, but unfortunately not for some edge cases:

1) ✔️ Press and hold A, then press, hold and release Shift → Works as expected, movement speeds up and slows down with the Shift key
2) ❌ Press and hold Shift-A, then release Shift while continuing to hold A → Starts with fast movement, on releasing Shift the movement stops for a moment and then resumes at a slower pace
3) ❌ Press and hold Shift-A, then release A while holding Shift, then press and hold A again (while still holding Shift) → Starts as fast movement, then stops, then resumes as slow movement

I've done some primitive debugging, by just logging what the TriggerListener sees:
For case 2, the trigger listener registers a begin-event for Shift-A and an end-event for both Shift and A, despite only releasing Shift.
Similarily, for case 3, the listener registers a begin and end event for Shift-A, followed by a begin event for just A.

I'm guessing this is because Shift-A is considered one event, given that Shift is a modifier, so it fails to register Shift as a separate key and always treats it in combination with the A key (but I would need to look into this further to confirm that).

Do you have any deeper insights on how the input handling works internally that might shed some more light on this?
I'd be happy to do some deeper debugging to figure out what's going on here.

**Reply by LegionaryCohort** [reply]
Quick follow-up:
Using the idea of trigger listeners, I've moved the logic for the A key into a second trigger listener (and consequently removed the UserActions completely).
This solves the second case, but the problem of the third case still remains, with the same events being recorded by the trigger listener.

If there is some way to decouple the Shift key, so that no end event is created for it when just releasing the A key, that should solve the issue. But I'm not sure whether that's possible.

**Reply by AlmasB** [reply]
Thanks, I'll give it some thought too. It's a common use case, so would be good to find a solution for it, whether external or internal.

**Reply by LegionaryCohort** [reply]
Looking a bit into the details of the Input class, it appears the root cause is the first line on the `handleReleased(event: InputEvent)` method:
`val releasedTriggers = activeTriggers.filter { it.isReleased(event) || (it is KeyTrigger && isIllegal(it.key)) }`
So basically whenever a key is released, it automatically also considers any of the `ILLEGAL_KEYS` (`CONTROL`, `SHIFT` and `ALT`) to have been released. This also applies if I'm pressing multiple other keys and release just one.

I've extended my example to include the A and W keys. If I press and hold Shift, A & W at the same time, I get fast movement to the top left corner (✔️), but if I release just the W key while continuing to hold Shift-A, I get slow movement to the left (❌).
The Shift key is effectively terminated, as suspected, when I release any one key of the ones I'm pressing.

**Reply by AlmasB** [reply]
Yes, you are correct, the right side of the check should read something like this: `it is KeyTrigger && it.key == (event as KeyEvent).code`. I'll turn this into an issue.

Doesn't solve your case fully I think, but it should fix this bug.

**Reply by AlmasB** [reply, marked as answer]
I've logged this as #1313 
Current tests are passing, so clearly this bug isn't covered. Feel free to write a unit test for this case if you have time, or just note some pseudo code here and I'll adapt.

**Reply by LegionaryCohort** [reply]
Not sure I have enough time to fit this in right this moment, but I'll look into it as soon as I can.

**Comment by LegionaryCohort** [comment]
I'll mark this as closed for now and continue further discussion in #1313
